File: API/tg_bots/Model/MenuItem.py
from typing import Optional

from pydantic import BaseModel, field_validator
from Model.Nutrients import Nutrients


class MenuItem(BaseModel):
    id: Optional[int]
    category_id: int
    cooking_time_minutes: int
    price: float
    name: str
    description: str
    image_path: Optional[str]
    nutrients_of_100_grams: Optional[Nutrients]

    def __init__(self,
                 category_id: int,
                 price: float,
                 name: str,
                 description: str,
                 cooking_time_minutes: int,
                 image_path: Optional[str] = None,
                 nutrients_of_100_grams: Optional[Nutrients] = None,
                 id: Optional[int] = None,
                 **kwargs):
        super().__init__(
            id=id,
            category_id=category_id,
            price=price,
            name=name,
            description=description,
            image_path=image_path,
            cooking_time_minutes=cooking_time_minutes,
            nutrients_of_100_grams=nutrients_of_100_grams,
            **kwargs
        )

    @field_validator('__all__', mode='before', check_fields=False)
    def determine_type(cls, values):
        from API.tg_bots.Model.Dish import Dish
        from API.tg_bots.Model.Drink import Drink

        if 'weight' in values:
            return Dish(**values)
        elif 'volume' in values:
            return Drink(**values)
        return values

File: Domain/Models/MenuItem.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Globalization;
using System.Text.Json.Serialization;
using System.Text.Json;
using Domain.DataBaseAccess;
using Microsoft.EntityFrameworkCore;

namespace Domain.Models;
[JsonConverter(typeof(MenuItemConverter))]
public class MenuItem : ITakeRelatedData, IDeleteRelatedData
{
    public long? Id { get; set; }
    [JsonPropertyName("category_id")] public long CategoryId { get; set; }
    [JsonPropertyName("cooking_time_minutes")] public double CookingTimeMinutes { get; set; }
    public double Price { get; set; }
    public string Name { get; set; }
    [JsonPropertyName("image_path")] public string? ImagePath { get; set; }
    public string Description { get; set; }
    [JsonPropertyName("nutrients_of_100_grams")] [NotMapped] public Nutrients? NutrientsOf100grams { get; set; }
    
    public MenuItem(){}
    public MenuItem(long? id, long categoryId, int cookingTimeMinutes, double price, string name, string description,
        string? imagePath, Nutrients? nutrientsOf100grams)
    {
        Id = id;
        CategoryId = categoryId;
        CookingTimeMinutes = cookingTimeMinutes;
        Price = price;
        Name = name;
        ImagePath = imagePath;
        Description = description;
        NutrientsOf100grams = nutrientsOf100grams;
    }

    public async Task TakeRelatedData(ApplicationDbContext context)
    {
        NutrientsOf100grams = await context.Nutrients.FirstOrDefaultAsync(n => n.MenuItemId == Id);
    }

    public void DeleteRelatedData(ApplicationDbContext context)
    {
        var nutrient = context.Nutrients.FirstOrDefault(n => n.MenuItemId == Id);
        if (nutrient == null) return;
        context.Nutrients.Remove(nutrient);
        context.SaveChanges();
    }

    public override string ToString()
    {
        var jsonObj = new
        {
            name = Name,
            image = ImagePath,
            price = Price,
            desc = Description,
            nutrients = NutrientsOf100grams ?? null,
            cooking_time = CookingTimeMinutes,
            volume = this is Dish ? ((Dish)this).Weight.ToString() + "г" : ((Drink)this).Volume.ToString() + "мл",
        };
        
        return JsonSerializer.Serialize(jsonObj);
    }
}